### PR TITLE
Implement Array.prototype.flat(Map) with some basic tests

### PR DESCRIPTION
**Pull request by MStoykov**

### CODE REVIEW

**Comment by MStoykov** [comment]
I will be making some more as [part of our try to drop corejs](https://github.com/loadimpact/k6/pull/1772) completely. 

The [list](https://github.com/loadimpact/k6/pull/1772#issuecomment-745417009) so far is :
- Array.prototype.flat(Map) - this PR
- String.prototype.trimLeft/Right 
- String.prototype.matchAll
- Object.prototype.getOwnPropertyDescriptors
- Object.prototype.values/entries 

Possibly (although I hope not):

- Object.prototype.__defineSetter/defineGetter__ and the lookup part

File: builtin_array.go [hunk 1]
```diff
@@ -1023,6 +1023,60 @@ func (r *Runtime) arrayproto_findIndex(call FunctionCall) Value {
...
+				element = mapperFunction(FunctionCall{
+					This:      thisArg,
+					Arguments: []Value{element, p, source},
+				})
+			}
+			if depth > 0 && isArray(element.ToObject(r)) {
+				elementLen := toLength(element.ToObject(r).self.getStr("length", nil))
+				targetIndex = r.flattenIntoArray(target, element.ToObject(r), elementLen, targetIndex, depth-1, mapperFunction, thisArg)
+			} else {
+				if targetIndex > maxInt {
```
Comment (dop251) [inline builtin_array.go:1054]:
```suggestion
				if targetIndex >= maxInt-1 {
```

File: builtin_array.go [hunk 2]
```diff
@@ -1023,6 +1023,60 @@ func (r *Runtime) arrayproto_findIndex(call FunctionCall) Value {
...
+					r.typeErrorResult(true, "Invalid array length")
+					panic("unreachable")
```
Comment (dop251) [inline builtin_array.go:1056]:
```suggestion
					panic(r.NewTypeError("Invalid array length"))
```

File: builtin_array.go [hunk 3]
```diff
@@ -1204,6 +1258,8 @@ func (r *Runtime) createArrayProto(val *Object) objectImpl {
 	o._putProp("filter", r.newNativeFunc(r.arrayproto_filter, nil, "filter", nil, 1), true, false, true)
 	o._putProp("find", r.newNativeFunc(r.arrayproto_find, nil, "find", nil, 1), true, false, true)
 	o._putProp("findIndex", r.newNativeFunc(r.arrayproto_findIndex, nil, "findIndex", nil, 1), true, false, true)
+	o._putProp("flat", r.newNativeFunc(r.arrayproto_flat, nil, "flat", nil, 0), true, false, true)
+	o._putProp("flatMap", r.newNativeFunc(r.arrayproto_flatMap, nil, "flatMap", nil, 1), true, false, true)
```
Comment (dop251) [inline builtin_array.go:1262]: Please also add these to unscopables here: https://github.com/dop251/goja/blob/27b0a7dc4c7fab42694a39c77da455a586da9173/builtin_array.go#L1242

**Review by dop251 — CHANGES_REQUESTED** [review]
Thanks for submitting. I have added a couple of minor changes, otherwise looks good to me.
The only thing is most array methods have optimisations for simple arrays (see `checkStdArray`) where property access can be replaced with operating directly on `a.values`. It's not necessary though and could be added at a later stage should there be a need...

File: builtin_array.go (hunk 2 above)
Comment (MStoykov) [reply builtin_array.go:1056]: I am fine with the change but just wanted to point out I copied it from 
https://github.com/dop251/goja/blob/51b43d63272ebd474a52ab0069fc510a8e1715e1/builtin_array.go#L114-L115

Comment (dop251) [reply builtin_array.go]: I know 😞 Thanks.

**Comment by MStoykov** [comment]
After the latest changes the only flat* test that are failing are :
```
--- FAIL: TestTC39 (4.22s)
    --- FAIL: TestTC39/test262 (4.22s)
        --- FAIL: TestTC39/test262/test/built-ins/Array/prototype/flat/null-undefined-elements.js (0.00s)
        Error Trace:    tc39_test.go:219
                        tc39_test.go:233
                        tc39_test.go:280
                        tc39_test.go:392
                        tc39_test.go:515
                        tc39_norace_test.go:11
        Error:          Should be empty, but was [test/built-ins/Array/prototype/flat/null-undefined-elements.js TypeError: Cannot convert undefined or null to object at flat (native)]: %!v(MISSING)
        Test:           TestTC39/test262/test/built-ins/Array/prototype/flat/null-undefined-elements.js
        --- FAIL: TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-bad-throws.js (0.45s)
        Error Trace:    tc39_test.go:219
                        tc39_test.go:233
                        tc39_test.go:280
                        tc39_test.go:392
                        tc39_test.go:515
                        tc39_norace_test.go:11
        Error:          Should be empty, but was [test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-bad-throws.js Test262Error: throw TypeError if @@species is a number Expected a TypeError to be thrown but no exception was thrown at all at harness/sta.js:22:9(49)]: %!v(MISSING)
        Test:           TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-bad-throws.js
        --- FAIL: TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor-poisoned-throws.js (0.17s)
        Error Trace:    tc39_test.go:219
                        tc39_test.go:233
                        tc39_test.go:280
                        tc39_test.go:392
                        tc39_test.go:515
                        tc39_norace_test.go:11
        Error:          Should be empty, but was [test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor-poisoned-throws.js Test262Error: Return abrupt completion from species custom ctor Expected a Test262Error to be thrown but no exception was thrown at all at harness/sta.js:22:9(49)]: %!v(MISSING)
        Test:           TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor-poisoned-throws.js
        --- FAIL: TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor.js (0.20s)
        Error Trace:    tc39_test.go:219
                        tc39_test.go:233
                        tc39_test.go:280
                        tc39_test.go:392
                        tc39_test.go:515
                        tc39_norace_test.go:11
        Error:          Should be empty, but was [test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor.js Test262Error: returned value is an instance of custom ctor at harness/sta.js:22:9(49)]: %!v(MISSING)
        Test:           TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species-custom-ctor.js
        --- FAIL: TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species.js (0.53s)
        Error Trace:    tc39_test.go:219
                        tc39_test.go:233
                        tc39_test.go:280
                        tc39_test.go:392
                        tc39_test.go:515
                        tc39_norace_test.go:11
        Error:          Should be empty, but was [test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species.js Test262Error: got species once Expected SameValue(«0», «1») to be true at harness/sta.js:22:9(49)]: %!v(MISSING)
        Test:           TestTC39/test262/test/built-ins/Array/prototype/flatMap/this-value-ctor-object-species.js
FAIL
exit status 1
```

For the record this was run through k6 which also employs babel to get some additional ES support including template variables. 


I don't think I can do much about any of those? 

but maybe I should now write the fast path? Including moving the check for the map function outside the loop and having 2 loops?

File: builtin_array.go [hunk 4]
```diff
@@ -1023,6 +1023,59 @@ func (r *Runtime) arrayproto_findIndex(call FunctionCall) Value {
...
+			if depth > 0 && isArray(element.ToObject(r)) {
+				elementLen := toLength(element.ToObject(r).self.getStr("length", nil))
+				targetIndex = r.flattenIntoArray(target, element.ToObject(r), elementLen, targetIndex, depth-1, nil, nil)
+			} else {
+				if targetIndex >= maxInt-1 {
+					panic(r.NewTypeError("Invalid array length"))
+				}
+				createDataPropertyOrThrow(target, intToValue(targetIndex), element)
+				targetIndex++
+			}
```
Comment (dop251) [inline builtin_array.go:1059]: `IsArray` should be applied directly to the element, so this block should be like this:
```suggestion
			var elementArray *Object
			if depth > 0 {
				if elementObj, ok := element.(*Object); ok && isArray(elementObj) {
					elementArray = elementObj
				}
			}
			if elementArray != nil {
				elementLen := toLength(elementArray.self.getStr("length", nil))
				targetIndex = r.flattenIntoArray(target, elementArray, elementLen, targetIndex, depth-1, nil, nil)
			} else {
				if targetIndex >= maxInt-1 {
					panic(r.NewTypeError("Invalid array length"))
				}
				createDataPropertyOrThrow(target, intToValue(targetIndex), element)
				targetIndex++
			}
```

**Review by dop251 — CHANGES_REQUESTED** [review]

File: builtin_array.go (hunk 4 above)
Comment (MStoykov) [reply builtin_array.go]: This did fix the tests with `undefined`
Thanks :)

Comment (dop251) [reply builtin_array.go]: As for the rest, I quickly backported and checked them, they all passed for me. I believe it has something to do with babel...

This looks good enough for merging. If you are going to implement any optimisations they may go as a separate PR(s)

**Review by dop251 — APPROVED** [review]

**Pull request merged**